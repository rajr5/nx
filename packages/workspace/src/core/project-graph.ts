export * from 'nx/src/core/project-graph/project-graph';
