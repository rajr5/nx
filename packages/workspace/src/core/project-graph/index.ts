export {
  createProjectGraph,
  createProjectGraphAsync,
  readCachedProjectGraph,
  createProjectGraphInCurrentProcess,
} from './project-graph';
export * from './project-graph-models';
export * from './operators';
