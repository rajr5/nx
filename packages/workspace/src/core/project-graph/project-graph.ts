import { mkdirSync, readFileSync } from 'fs';
import { ProjectGraph, ProjectGraphContext } from './project-graph-models';
import { ProjectGraphBuilder } from './project-graph-builder';
import { appRootPath } from '../../utils/app-root';
import {
  directoryExists,
  fileExists,
  readJsonFile,
  writeJsonFile
} from '../../utils/fileutils';
import {
  FileData,
  mtime,
  readNxJson,
  readWorkspaceFiles,
  readWorkspaceJson
} from '../file-utils';
import { createFileMap, FileMap } from '../file-graph';
import { BuildNodes, buildWorkspaceProjectNodes } from './build-nodes';
import {
  BuildDependencies,
  buildExplicitTypeScriptDependencies,
  buildImplicitProjectDependencies
} from './build-dependencies';
import { assertWorkspaceValidity } from '../assert-workspace-validity';

export function createProjectGraph(
  workspaceJson = readWorkspaceJson(),
  nxJson = readNxJson(),
  workspaceFiles = readWorkspaceFiles(),
  fileRead: (s: string) => string = defaultFileRead,
  cache: false | { data: ProjectGraphCache; mtime: number } = readCache()
): ProjectGraph {
  assertWorkspaceValidity(workspaceJson, nxJson);

  if (!cache || maxMTime(workspaceFiles) > cache.mtime) {
    const builder = new ProjectGraphBuilder(
      cache ? cache.data.projectGraph : undefined
    );
    const buildNodesFns: BuildNodes[] = [buildWorkspaceProjectNodes];
    const buildDependenciesFns: BuildDependencies[] = [
      buildExplicitTypeScriptDependencies,
      buildImplicitProjectDependencies
    ];
    // TODO: File graph needs to be handled separately from project graph, or
    // we need a way to support project extensions (e.g. npm project).
    const fileMap = createFileMap(workspaceJson, workspaceFiles);
    const ctx = {
      workspaceJson,
      nxJson,
      fileMap: modifiedSinceCache(fileMap, cache)
    };

    buildNodesFns.forEach(f => f(ctx, builder.addNode.bind(builder), fileRead));

    buildDependenciesFns.forEach(f =>
      f(ctx, builder.nodes, builder.addDependency.bind(builder), fileRead)
    );

    const projectGraph = builder.build();
    writeCache({
      projectGraph,
      fileMap
    });
    return projectGraph;
  } else {
    // Cache file was modified _after_ all workspace files.
    // Safe to return the cached graph.
    return cache.data.projectGraph;
  }
}

// -----------------------------------------------------------------------------

interface ProjectGraphCache {
  projectGraph: ProjectGraph;
  fileMap: FileMap;
}

function defaultFileRead(filePath: string) {
  return readFileSync(`${appRootPath}/${filePath}`, 'UTF-8');
}

const nxDepsPath = `${appRootPath}/dist/nxdeps.json`;

function readCache(): false | { data: ProjectGraphCache; mtime: number } {
  if (!directoryExists(`${appRootPath}/dist`)) {
    mkdirSync(`${appRootPath}/dist`);
  }

  const data = getValidCache(
    fileExists(nxDepsPath) ? readJsonFile(nxDepsPath) : null
  );

  return data ? { data, mtime: mtime(nxDepsPath) } : false;
}

function getValidCache(cache: ProjectGraphCache | null) {
  if (!cache) {
    return null;
  }
  return cache.projectGraph && cache.fileMap ? cache : null;
}

function writeCache(cache: ProjectGraphCache): void {
  writeJsonFile(nxDepsPath, cache);
}

function maxMTime(files: FileData[]) {
  return Math.max(...files.map(f => f.mtime));
}

function modifiedSinceCache(
  fileMap: FileMap,
  c: false | { data: ProjectGraphCache; mtime: number }
): FileMap {
  // No cache -> compute entire graph
  if (!c) {
    return fileMap;
  }

  const currentProjects = Object.keys(fileMap).sort();
  const previousProjects = Object.keys(c.data.fileMap).sort();

  // Projects changed -> compute entire graph
  if (
    currentProjects.length !== previousProjects.length ||
    currentProjects.some((val, idx) => val !== previousProjects[idx])
  ) {
    return fileMap;
  }

  // Projects are same, only compute changed projects
  const modifiedSince: FileMap = currentProjects.reduce((acc, k) => {
    acc[k] = [];
    return acc;
  }, {});
  currentProjects.forEach(p => {
    for (const f of fileMap[p]) {
      const fromCache = c.data.fileMap[p].find(x => x.file === f.file);

      if (!fromCache) {
        modifiedSince[p].push(f);
      } else if (f.mtime > fromCache.mtime) {
        modifiedSince[p].push(f);
      }
    }
  });
  return modifiedSince;
}
