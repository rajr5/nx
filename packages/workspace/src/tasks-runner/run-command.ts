import {
  AffectedEventType,
  Task,
  TaskCompleteEvent,
  TasksRunner
} from './tasks-runner';
import { defaultTasksRunner } from './default-tasks-runner';
import { isRelativePath } from '../utils/fileutils';
import { join } from 'path';
import { appRootPath } from '../utils/app-root';
import { DefaultReporter, ReporterArgs } from './default-reporter';
import * as yargs from 'yargs';
import { ProjectGraph, ProjectGraphNode } from '../core/project-graph';
import { Environment, NxJson } from '../core/shared-interfaces';
import { projectHasTargetAndConfiguration } from '../utils/project-has-target-and-configuration';
import { NxArgs } from '@nrwl/workspace/src/command-line/utils';

export interface TasksMap {
  [projectName: string]: { [targetName: string]: Task };
}

type RunArgs = yargs.Arguments & ReporterArgs;

export function runCommand<T extends RunArgs>(
  projectsToRun: ProjectGraphNode[],
  projectGraph: ProjectGraph,
  { nxJson, workspace }: Environment,
  nxArgs: NxArgs,
  overrides: any
) {
  const reporter = new DefaultReporter();
  reporter.beforeRun(projectsToRun.map(p => p.name), nxArgs, overrides);
  const tasks: Task[] = projectsToRun.map(project =>
    createTask({
      project,
      target: nxArgs.target,
      configuration: nxArgs.configuration,
      overrides: overrides
    })
  );

  const tasksMap: TasksMap = {};
  Object.entries(projectGraph.nodes).forEach(([projectName, project]) => {
    const runnable = projectHasTargetAndConfiguration(
      project,
      nxArgs.target,
      nxArgs.configuration
    );
    if (runnable) {
      tasksMap[projectName] = {
        [nxArgs.target]: createTask({
          project: project,
          target: nxArgs.target,
          configuration: nxArgs.configuration,
          overrides: overrides
        })
      };
    }
  });

  const { tasksRunner, tasksOptions } = getRunner(
    nxArgs.runner,
    nxJson,
    overrides
  );
  tasksRunner(tasks, tasksOptions, {
    target: nxArgs.target,
    projectGraph,
    tasksMap
  }).subscribe({
    next: (event: TaskCompleteEvent) => {
      switch (event.type) {
        case AffectedEventType.TaskComplete: {
          workspace.setResult(event.task.target.project, event.success);
        }
      }
    },
    error: console.error,
    complete: () => {
      // fix for https://github.com/nrwl/nx/issues/1666
      if (process.stdin['unref']) (process.stdin as any).unref();

      workspace.saveResults();
      reporter.printResults(
        nxArgs,
        workspace.failedProjects,
        workspace.startedWithFailedProjects
      );

      if (workspace.hasFailure) {
        process.exit(1);
      }
    }
  });
}

export interface TaskParams {
  project: ProjectGraphNode;
  target: string;
  configuration: string;
  overrides: Object;
}

export function createTask({
  project,
  target,
  configuration,
  overrides
}: TaskParams): Task {
  return {
    id: getId({
      project: project.name,
      target: target,
      configuration: configuration
    }),
    target: {
      project: project.name,
      target,
      configuration
    },
    overrides: interpolateOverrides(overrides, project.name, project.data)
  };
}

function getId({
  project,
  target,
  configuration
}: {
  project: string;
  target: string;
  configuration?: string;
}): string {
  let id = project + ':' + target;
  if (configuration) {
    id += ':' + configuration;
  }
  return id;
}

export function getRunner(
  runner: string | undefined,
  nxJson: NxJson,
  overrides: any
): {
  tasksRunner: TasksRunner;
  tasksOptions: unknown;
} {
  if (!nxJson.tasksRunnerOptions) {
    return {
      tasksRunner: defaultTasksRunner,
      tasksOptions: overrides
    };
  }

  if (!runner && !nxJson.tasksRunnerOptions.default) {
    return {
      tasksRunner: defaultTasksRunner,
      tasksOptions: overrides
    };
  }

  runner = runner || 'default';

  if (nxJson.tasksRunnerOptions[runner]) {
    let modulePath: string = nxJson.tasksRunnerOptions[runner].runner;
    if (isRelativePath(modulePath)) {
      modulePath = join(appRootPath, modulePath);
    }

    let tasksRunner = require(modulePath);
    // to support both babel and ts formats
    if (tasksRunner.default) {
      throw new Error('boom');
      tasksRunner = tasksRunner.default;
    }

    return {
      tasksRunner,
      tasksOptions: {
        ...nxJson.tasksRunnerOptions[runner].options,
        ...overrides
      }
    };
  } else {
    throw new Error(`Could not find runner configuration for ${runner}`);
  }
}

function interpolateOverrides<T = any>(
  args: T,
  projectName: string,
  projectMetadata: any
): T {
  const interpolatedArgs: T = { ...args };
  Object.entries(interpolatedArgs).forEach(([name, value]) => {
    if (typeof value === 'string') {
      const regex = /{project\.([^}]+)}/g;
      interpolatedArgs[name] = value.replace(regex, (_, group: string) => {
        if (group.includes('.')) {
          throw new Error('Only top-level properties can be interpolated');
        }

        if (group === 'name') {
          return projectName;
        }
        return projectMetadata[group];
      });
    }
  });
  return interpolatedArgs;
}
