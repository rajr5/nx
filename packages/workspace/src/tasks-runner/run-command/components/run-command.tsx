import { Text, useApp } from 'ink';
import * as React from 'react';
import { useState } from 'react';
import { TaskResult } from '../../life-cycle';
import { Task } from '../../tasks-runner';
import { InkRunManyLifeCycle } from '../ink-run-many-life-cycle';
import { NoTargetsRun } from './no-targets-run';
import { RunMany } from './run-many';

type RunCommandState = 'INIT' | 'NO_TARGETS_RUN' | 'RUN_MANY';

export interface TasksState {
  target: string;
  projectNames: string[];
  tasks: Task[];
  taskResults: TaskResult[];
  tasksToTerminalOutputs: { [taskId: string]: string };
}

export function RunCommandComponent({
  lifeCycle,
}: {
  lifeCycle: InkRunManyLifeCycle;
}) {
  const [runCommandState, setRunCommandState] =
    useState<RunCommandState>('INIT');
  const [onStartCommandParamsState, setOnStartCommandParamsState] =
    useState(null);
  const [tasksState, setTasksState] = useState<TasksState | null>(null);
  const { exit } = useApp();

  lifeCycle.callbacks.onStartCommand = (params) => {
    setOnStartCommandParamsState(params);
    if (params.projectNames.length <= 0) {
      setRunCommandState('NO_TARGETS_RUN');
      return exit();
    }
    setTasksState({
      target: params.args.target,
      projectNames: params.projectNames,
      tasks: null,
      taskResults: null,
      tasksToTerminalOutputs: {},
    });
    setRunCommandState('RUN_MANY');
  };

  lifeCycle.callbacks.onStartTasks = (tasks) => {
    setTasksState((state) => ({
      ...state,
      tasks: [...(state.tasks || []), ...tasks],
    }));
  };

  lifeCycle.callbacks.onEndTasks = (taskResults) => {
    setTasksState((state) => ({
      ...state,
      taskResults: [...(state.taskResults || []), ...taskResults],
    }));
  };

  lifeCycle.callbacks.onPrintTaskTerminalOutput = (
    task,
    _cacheStatus,
    output
  ) => {
    setTasksState((state) => ({
      ...state,
      tasksToTerminalOutputs: {
        ...state.tasksToTerminalOutputs,
        [task.id]: output,
      },
    }));
  };

  function clearTaskOutputCacheForTaskId(taskId: string) {
    setTasksState((state) => {
      const updatedTasksToTerminalOutputs = {
        ...state.tasksToTerminalOutputs,
      };
      delete updatedTasksToTerminalOutputs[taskId];
      return {
        ...state,
        tasksToTerminalOutputs: updatedTasksToTerminalOutputs,
      };
    });
  }

  switch (runCommandState) {
    case 'NO_TARGETS_RUN':
      return <NoTargetsRun onStartCommandParams={onStartCommandParamsState} />;
    case 'RUN_MANY':
      return (
        <RunMany
          tasksState={tasksState}
          clearTaskOutputCacheForTaskId={clearTaskOutputCacheForTaskId}
        />
      );
    default:
      return <Text>{''}</Text>;
  }
}
