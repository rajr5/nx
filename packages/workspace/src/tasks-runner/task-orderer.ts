import { ProjectGraph } from '../core/project-graph';
import { Task } from './tasks-runner';
import { DefaultTasksRunnerOptions } from './default-tasks-runner';
import { getDependencyConfigs } from './utils';
import { performance } from 'perf_hooks';
import { TargetDependencyConfig } from '@nrwl/tao/src/shared/workspace';

interface TaskGraph {
  tasks: Record<string, Task>;
  dependencies: Record<string, string[]>;
}

export class TaskOrderer {
  constructor(
    private readonly projectGraph: ProjectGraph,
    private readonly defaultTargetDependencies: Record<
      string,
      TargetDependencyConfig[]
    >
  ) {}

  splitTasksIntoStages(tasks: Task[]) {
    if (tasks.length === 0) return [];

    const stages: Task[][] = [];
    performance.mark('ordering tasks:start');
    const taskGraph = this.createTaskGraph(tasks);
    const notStagedTaskIds = new Set<string>(tasks.map((t) => t.id));
    let stageIndex = 0;

    // Loop through tasks and try to stage them. As tasks are staged, they are removed from the loop
    while (notStagedTaskIds.size > 0) {
      const currentStage = (stages[stageIndex] = []);
      for (const taskId of notStagedTaskIds) {
        let ready = true;
        for (const dependency of taskGraph.dependencies[taskId]) {
          if (notStagedTaskIds.has(dependency)) {
            // dependency has not been staged yet, this task is not ready to be staged.
            ready = false;
            break;
          }
        }

        // Some dependency still has not been staged, skip it for now, it will be processed again
        if (!ready) {
          continue;
        }

        // All the dependencies have been staged, let's stage it.
        const task = taskGraph.tasks[taskId];
        currentStage.push(task);
      }

      // Remove the entire new stage of tasks from the list
      for (const task of currentStage) {
        notStagedTaskIds.delete(task.id);
      }
      stageIndex++;
    }
    performance.mark('ordering tasks:end');
    performance.measure(
      'ordering tasks',
      'ordering tasks:start',
      'ordering tasks:end'
    );
    return stages;
  }

  private createTaskGraph(tasks: Task[]): TaskGraph {
    const graph: TaskGraph = {
      tasks: {},
      dependencies: {},
    };
    for (const task of tasks) {
      graph.tasks[task.id] = task;
      graph.dependencies[task.id] = [];
      const dependencyConfigs = getDependencyConfigs(
        task.target,
        this.defaultTargetDependencies,
        this.projectGraph
      );

      if (!dependencyConfigs) {
        continue;
      }

      const projectDependencies = new Set(
        this.projectGraph.dependencies[task.target.project].map(
          (dependency) => dependency.target
        )
      );

      for (const dependencyConfig of dependencyConfigs) {
        if (dependencyConfig.projects === 'self') {
          for (const t of tasks) {
            if (
              t.target.project === task.target.project &&
              t.target.target === dependencyConfig.target
            ) {
              graph.dependencies[task.id].push(t.id);
            }
          }
        } else if (dependencyConfig.projects === 'dependencies') {
          for (const t of tasks) {
            if (
              projectDependencies.has(t.target.project) &&
              t.target.target === dependencyConfig.target
            ) {
              graph.dependencies[task.id].push(t.id);
            }
          }
        }
      }
    }
    return graph;
  }
}
