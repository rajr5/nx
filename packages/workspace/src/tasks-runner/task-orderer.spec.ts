import { TaskOrderer } from './task-orderer';
import { DependencyType } from '../core/project-graph';
import { createTasksForProjectToRun } from './run-command';
import { ProjectGraph } from '@nrwl/devkit';

describe('TaskStages', () => {
  it('should return empty for an empty array', () => {
    const stages = new TaskOrderer({} as any, 'build', {
      nodes: {},
      dependencies: {},
    }).splitTasksIntoStages([]);
    expect(stages).toEqual([]);
  });

  it('should split tasks into stages based on their dependencies', () => {
    const stages = new TaskOrderer({} as any, 'build', {
      nodes: {
        child1: { type: 'lib' },
        child2: { type: 'lib' },
        parent: { type: 'lib' },
        grandparent: { type: 'lib' },
      } as any,
      dependencies: {
        child1: [],
        child2: [],
        parent: [
          {
            source: 'parent',
            target: 'child1',
            type: DependencyType.static,
          },
          {
            source: 'parent',
            target: 'child2',
            type: DependencyType.static,
          },
        ],
        grandparent: [
          {
            source: 'grandparent',
            target: 'parent',
            type: DependencyType.static,
          },
        ],
      },
    }).splitTasksIntoStages([
      {
        id: 'parent:build',
        target: { project: 'parent', target: 'build' },
      },
      {
        id: 'child1:build',
        target: { project: 'child1', target: 'build' },
      },
      {
        id: 'child2:build',
        target: { project: 'child2', target: 'build' },
      },
      {
        id: 'grandparent:build',
        target: { project: 'grandparent', target: 'build' },
      },
    ] as any);

    expect(stages).toEqual([
      [
        {
          id: 'child1:build',
          target: { project: 'child1', target: 'build' },
        },
        {
          id: 'child2:build',
          target: { project: 'child2', target: 'build' },
        },
      ],
      [
        {
          id: 'parent:build',
          target: { project: 'parent', target: 'build' },
        },
      ],
      [
        {
          id: 'grandparent:build',
          target: { project: 'grandparent', target: 'build' },
        },
      ],
    ]);
  });

  it('should support custom targets that require strict ordering', () => {
    const stages = new TaskOrderer(
      { strictlyOrderedTargets: ['custom'] } as any,
      'custom',
      {
        nodes: { child1: { type: 'lib' }, parent: { type: 'lib' } } as any,
        dependencies: {
          child1: [],
          parent: [
            {
              source: 'parent',
              target: 'child1',
              type: DependencyType.static,
            },
          ],
        },
      }
    ).splitTasksIntoStages([
      {
        id: 'parent:custom',
        target: { project: 'parent', target: 'custom' },
      },
      {
        id: 'child1:custom',
        target: { project: 'child1', target: 'custom' },
      },
    ] as any);

    expect(stages).toEqual([
      [
        {
          id: 'child1:custom',
          target: { project: 'child1', target: 'custom' },
        },
      ],
      [
        {
          id: 'parent:custom',
          target: { project: 'parent', target: 'custom' },
        },
      ],
    ]);

    const noStages = new TaskOrderer(
      { strictlyOrderedTargets: ['custom'] } as any,
      'some-other-custom',
      {
        nodes: {},
        dependencies: {
          child1: [],
          parent: [
            {
              source: 'parent',
              target: 'child1',
              type: DependencyType.static,
            },
          ],
        },
      }
    ).splitTasksIntoStages([
      {
        target: { project: 'parent' },
      },
      {
        target: { project: 'child1' },
      },
    ] as any);

    expect(noStages).toEqual([
      [
        {
          target: { project: 'parent' },
        },
        {
          target: { project: 'child1' },
        },
      ],
    ]);
  });

  it('should split tasks into stages based on their dependencies when there are unrelated packages', () => {
    const stages = new TaskOrderer({} as any, 'build', {
      nodes: {
        app1: { type: 'lib' },
        app2: { type: 'lib' },
        common1: { type: 'lib' },
        common2: { type: 'lib' },
      } as any,
      dependencies: {
        app1: [
          {
            source: 'app1',
            target: 'common1',
            type: DependencyType.static,
          },
        ],
        app2: [
          {
            source: 'app2',
            target: 'common2',
            type: DependencyType.static,
          },
        ],
        common1: [],
        common2: [],
      },
    }).splitTasksIntoStages([
      {
        id: 'app1:build',
        target: { project: 'app1', target: 'build' },
      },
      {
        id: 'app2:build',
        target: { project: 'app2', target: 'build' },
      },
      {
        id: 'common1:build',
        target: { project: 'common1', target: 'build' },
      },
      {
        id: 'common2:build',
        target: { project: 'common2', target: 'build' },
      },
    ] as any);

    expect(stages).toEqual([
      [
        {
          id: 'common1:build',
          target: { project: 'common1', target: 'build' },
        },
        {
          id: 'common2:build',
          target: { project: 'common2', target: 'build' },
        },
      ],
      [
        {
          id: 'app1:build',
          target: { project: 'app1', target: 'build' },
        },
        {
          id: 'app2:build',
          target: { project: 'app2', target: 'build' },
        },
      ],
    ]);
  });

  describe('(tasks with dependency configurations)', () => {
    let projectGraph: ProjectGraph;
    beforeEach(() => {
      projectGraph = {
        nodes: {
          app1: {
            type: 'app',
            name: 'app1',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'app1-root',
              files: [],
            },
          },
          app2: {
            type: 'app',
            name: 'app2',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'app2-root',
              files: [],
            },
          },
          common1: {
            type: 'lib',
            name: 'common1',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'common1-root',
              files: [],
            },
          },
          common2: {
            type: 'lib',
            name: 'common2',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'common2-root',
              files: [],
            },
          },
        },
        dependencies: {
          app1: [
            {
              source: 'app1',
              target: 'common1',
              type: DependencyType.static,
            },
          ],
          app2: [
            {
              source: 'app2',
              target: 'common2',
              type: DependencyType.static,
            },
          ],
          common1: [],
          common2: [],
        },
      };
    });

    it('should split tasks into stages (builds depend on build of dependencies)', () => {
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const stages = new TaskOrderer(
        {} as any,
        'build',
        projectGraph
      ).splitTasksIntoStages(tasks);

      // lib builds should be in their own stage
      // app builds should be in a later stage
      expect(stages).toMatchSnapshot();
    });

    it('should split tasks into stages  (builds depend on build of dependencies and prebuild of self)', () => {
      projectGraph.nodes.app1.data.targets.prebuild = {};
      projectGraph.nodes.app2.data.targets.prebuild = {};
      projectGraph.nodes.app1.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      projectGraph.nodes.app2.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );
      const stages = new TaskOrderer(
        {} as any,
        'build',
        projectGraph
      ).splitTasksIntoStages(tasks);

      // lib builds with the prebuild tasks in a stage
      // app builds should be in a later stage
      expect(stages).toMatchSnapshot();
    });

    it('should split tasks into stages  (builds depend on build of dependencies and prebuild of self, prebuilds depends on build of dependencies)', () => {
      projectGraph.nodes.app1.data.targets.prebuild = {
        dependsOn: [
          {
            target: 'build',
            projects: 'dependencies',
          },
        ],
      };
      projectGraph.nodes.app2.data.targets.prebuild = {
        dependsOn: [
          {
            target: 'build',
            projects: 'dependencies',
          },
        ],
      };
      projectGraph.nodes.app1.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      projectGraph.nodes.app2.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );
      const stages = new TaskOrderer(
        {} as any,
        'build',
        projectGraph
      ).splitTasksIntoStages(tasks);

      // lib builds in the first stage
      // app prebuilds in the next stage
      // app builds in the last stage
      expect(stages).toMatchSnapshot();
    });

    it('should split tasks into stages  (builds depend on build of dependencies, builds depend on prebuilds)', () => {
      projectGraph.nodes.common1.data.targets = projectGraph.nodes.common2.data.targets = {
        prebuild: {},
        build: {
          dependsOn: [
            {
              target: 'build',
              projects: 'dependencies',
            },
            {
              target: 'prebuild',
              projects: 'self',
            },
          ],
        },
      };
      projectGraph.nodes.app1.data.targets = projectGraph.nodes.app2.data.targets = {
        build: {
          dependsOn: [
            {
              target: 'build',
              projects: 'dependencies',
            },
          ],
        },
      };
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const stages = new TaskOrderer(
        {} as any,
        'build',
        projectGraph
      ).splitTasksIntoStages(tasks);

      // lib prebuilds in the first stage
      // lib builds in the next stage
      // app builds in the last stage
      expect(stages).toMatchSnapshot();
    });
  });
});
