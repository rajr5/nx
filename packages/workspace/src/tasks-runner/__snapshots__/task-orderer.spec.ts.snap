// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskStages (tasks with dependency configurations) should split tasks into stages  (builds depend on build of dependencies and prebuild of self) 1`] = `
Array [
  Array [
    Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    Object {
      "id": "app1:prebuild",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "prebuild",
      },
    },
    Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
    Object {
      "id": "app2:prebuild",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "prebuild",
      },
    },
  ],
  Array [
    Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
  ],
]
`;

exports[`TaskStages (tasks with dependency configurations) should split tasks into stages  (builds depend on build of dependencies and prebuild of self, prebuilds depends on build of dependencies) 1`] = `
Array [
  Array [
    Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  ],
  Array [
    Object {
      "id": "app1:prebuild",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "prebuild",
      },
    },
    Object {
      "id": "app2:prebuild",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "prebuild",
      },
    },
  ],
  Array [
    Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
  ],
]
`;

exports[`TaskStages (tasks with dependency configurations) should split tasks into stages  (builds depend on build of dependencies, builds depend on prebuilds) 1`] = `
Array [
  Array [
    Object {
      "id": "common1:prebuild",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "prebuild",
      },
    },
    Object {
      "id": "common2:prebuild",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "prebuild",
      },
    },
  ],
  Array [
    Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  ],
  Array [
    Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
  ],
]
`;

exports[`TaskStages (tasks with dependency configurations) should split tasks into stages (builds depend on build of dependencies) 1`] = `
Array [
  Array [
    Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  ],
  Array [
    Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
  ],
]
`;
